### ISSUE TITLE
Support for half-precision complex numbers?

### ISSUE THREAD

**Issue by leofang** [cupy/cupy#3370]
This issue is mainly for tracking the upstream counterpart (https://github.com/numpy/numpy/issues/14753) for compatibility purposes. However, there are unique challenges and opportunities in CUDA that makes this support worth considering independently of upstream (for the moment). 

Specifically, many scientific computing problems may have a portion (could be a significant one) that just needs low precision. For these applications, using half complex could further save both memory footprint and computing time. However, most of current CUDA libraries do not support half complex, so in CuPy a type cast to single complex is a must if CUDA libraries are to be used. Using elementwise/reduction kernels is free of such a problem, though, so it'd be interesting to target them as the first step.

If I were to work on this, I would start from investigating whether `thrust::complex` can support `complex<__half>` or not, and build up the infrastructure gradually.

cc: @smarkesini

**Comment by leofang** [comment]
@anaruse @maxpkatz Just curious, has NVIDIA considered providing support for half complex?

**Comment by kmaehashi** [comment]
Another thing to consider is that we need to define complex32 by our own as currently cupy borrow all dtypes from numpy.

**Comment by leofang** [comment]
Right. It might be easier if we unify the dtype access. Currently both namespaces (e.g., `cupy.float64` vs `numpy.float64`) are used in the codebase. If we solely use the former, we could then implement `cupy.complex32` and make it used across the realm. Once NumPy catches up, we then just remove our implementation and alias to `numpy.complex32`.

**Comment by Vlad0922** [comment]
As some pros of adding complex32 dtype; on modern NVidia architectures with tensor cores, operations with float16 are faster comparing to float32. So complex32 should also be faster in comparison with complex64.

**Comment by anaruse** [comment]
You can get a better matrix multiply performance with TensorCore, but fp16 complex numbers are not supported by CUDA libraries except [cublasLtMatmul()](https://docs.nvidia.com/cuda/cublas/index.html#cublasLtMatmul) in cuBLASTLt. Even with cublasLtMatmul(), all matrices must be in planar complex format, so format conversion will be needed before and after the call.

**Comment by Vlad0922** [comment]
Didnt know about that, thanks for the info.

**Comment by carterbox** [comment]
> The cudaDataType data type is an enumeration of the types supported by CUDA libraries.
> ```
> typedef enum cudaDataType_t
> {
>         CUDA_R_16F= 2, // 16 bit real 
>         CUDA_C_16F= 6, // 16 bit complex
>         CUDA_R_32F= 0, // 32 bit real
>         CUDA_C_32F= 4, // 32 bit complex
>         CUDA_R_64F= 1, // 64 bit real
>         CUDA_C_64F= 5, // 64 bit complex
>         CUDA_R_8I= 3,  // 8 bit real as a signed integer 
>         CUDA_C_8I= 7,  // 8 bit complex as a pair of signed integers
>         CUDA_R_8U= 8,  // 8 bit real as a signed integer 
>         CUDA_C_8U= 9   // 8 bit complex as a pair of signed integers
> } cudaDataType;
> ```

It looks like the [CUDA documentation](https://docs.nvidia.com/cuda/cufft/index.html#common-cudadatatype-types) mentions complex half-precision as far back as toolkit 8 as part of the `cudaDataType` enum. This enum is mentioned in the CuSOLVER, CuBLAS, CuFFT, and CuSPARSE documentation.

Has anyone actually tried using the half-complex? The documentation seems to indicate that there is broad support for this type.

@anaruse @leofang

**Comment by leofang** [comment]
Hi @carterbox I checked cuFFT and cuBLAS quickly:
- cuFFT: It seems possible to do C2C/R2C/C2R transforms involving `complex32` if we use the `cufftXtMakePlanMany()` API to create plans. I never test it myself, but I could add wrappers to them easily for further evaluation in the Python space.
- cuBLAS: Note that only a very limited subset of cuBLAS functions (suffixed with "Ex") accept `cudaDataType`. A few of them are already added to CuPy (ex: `cublasGemmEx`), but the coverage is far from complete.

I could take a look at other CUDA libraries later. @anaruse knows far better than I do so I'm sure he can share more 😁 IMHO, though, I think it's easier if we settle the infrastructure issues first, likely in the following order: 
1. Implement a new dtype `cupy.complex32` and make it recognized across the CuPy realm (https://github.com/cupy/cupy/issues/3370#issuecomment-637363494)
2. Ensure `thrust::complex<__half>` works at the C++ level (https://github.com/cupy/cupy/issues/3370#issue-625345215)
3. Support `cupy.complex32` in CuPy's ufuncs and reduction kernels (https://github.com/cupy/cupy/issues/3370#issue-625345215)
4. Make the test helpers in `cupy.testing` recognize `cupy.complex32`

Then we worry about incorporating the CUDA libraries (except for cuFFT, which might be an easier target).

**Comment by leofang** [comment]
Another thing is how to make this work in HIP...😞

**Comment by leofang** [comment]
Just a little update: I think most of the Thrust complex headers (for which CuPy keeps its own copy) should just work:
```python
import cupy as cp


code = r'''
#include <cupy/complex.cuh>  // for complex<T>
#include <cupy/carray.cuh>  // for float16


extern "C" __global__ void my_arange(complex<float16>* a, int N) {
    unsigned int tid = threadIdx.x + blockIdx.x * blockDim.x;

    if (tid < N) {
        a[tid] = complex<float16>((float16)(tid), 2*(float16)(tid));
    }
}
'''

ker = cp.RawKernel(code, 'my_arange', options=('-std=c++11',))
N = 10
arr = cp.empty(2*N, dtype=cp.float16)
ker((1,), (N,), (arr, N))
print(arr)  # [ 0.  0.  1.  2.  2.  4.  3.  6.  4.  8.  5. 10.  6. 12.  7. 14.  8. 16. 9. 18.]
```
For the constructors and arithmetic operations there is no problem at all. For math function calls like `sin`, `cos`, etc, which would be used in ufuncs, there would be an implicit type conversion to (from) single precision before (after) the computation, since `float16` defines this behavior. This may not be desired in the long run, and we will need Thrust people (cc: @allisonvacanti 😅) to handle this part as it requires serious numerical analysis to correctly sort everything out, but to get things up and running this is not a blocker.

**Comment by allisonvacanti** [comment]
I should point out that thrust will eventually be replacing its `complex` header with the libcu++ (https://github.com/NVIDIA/libcudacxx) implementation, so keep an eye on how that effort evolves.

**Comment by leofang** [comment]
> I should point out that thrust will eventually be replacing its `complex` header with the libcu++ (https://github.com/NVIDIA/libcudacxx) implementation, so keep an eye on how that effort evolves.

Thanks, @allisonvacanti! It's good to know. So does libcu++ already support `complex<__half>`? Would `thrust::complex` be removed completely, or become an alias of `cuda::std::complex`?

**Comment by allisonvacanti** [comment]
`__half` support has been discussed, but I don't know the current plans are or how it would be implemented. Feel free to file an issue on the libcudacxx repo to let them know you're interested in it.

I'm not sure exactly how we'll handle the switch for thrust yet, it'll depend on how compatible the implementations are.

**Comment by leofang** [comment]
> `__half` support has been discussed, but I don't know the current plans are or how it would be implemented. Feel free to file an issue on the libcudacxx repo to let them know you're interested in it.

Thanks! See NVIDIA/libcudacxx#95.

**Comment by jrhemstad** [comment]
fyi, `cuda::std::complex` support for `__half` and `bfloat` will be merged into CCCL soon: https://github.com/NVIDIA/cccl/issues/1139

**Comment by jakirkham** [comment]
> fyi, `cuda::std::complex` support for `__half` and `bfloat` will be merged into CCCL soon: [NVIDIA/cccl#1139](https://github.com/NVIDIA/cccl/issues/1139)

AIUI [this is in CCCL 2.4.0+]( https://github.com/NVIDIA/cccl/blob/v2.4.0/libcudacxx/docs/standard_api/numerics_library/complex.md#support-for-half-and-bfloat16-since-libcu-240 )

**Comment by seberg** [comment]
`mldtypes == 0.6.0` now ships `complex32` and `bcomplex32` (with more improvements coming).  This means that basic support for both of these dtypes is now effectively unlocked.
